### PR TITLE
Benchmarks started

### PR DESCRIPTION
**Pull request by odeke-em**
Trace Benchmarks

Updates #111

Work in progress to start benchmarks. Containing both
* a REST ping-pong server
* gRPC ping-pong server

traced vs untraced

giving such results:

* gRPC ping-pong server
```shell
$ go test -bench=. ./...
goos: darwin
goarch: amd64
pkg: go.opencensus.io/benchmarks/trace/grpc
BenchmarkTracedPong2QPS-4        	       1	1002930235 ns/op	  207640 B/op	     841 allocs/op
BenchmarkTracedPong10QPS-4       	       1	1004424371 ns/op	  361312 B/op	    1490 allocs/op
BenchmarkTracedPong100QPS-4      	       1	1001152449 ns/op	  739776 B/op	   11177 allocs/op
BenchmarkTracedPong1000QPS-4     	       1	1000633176 ns/op	 6038656 B/op	  108609 allocs/op
BenchmarkTracedPong5000QPS-4     	       1	1014634531 ns/op	29291216 B/op	  540696 allocs/op
BenchmarkUnTracedPong2QPS-4      	       1	1004254675 ns/op	  718208 B/op	     914 allocs/op
BenchmarkUnTracedPong10QPS-4     	       1	1002483956 ns/op	  210704 B/op	    1338 allocs/op
BenchmarkUnTracedPong100QPS-4    	       1	1001075685 ns/op	  849536 B/op	   10954 allocs/op
BenchmarkUnTracedPong1000QPS-4   	       1	1000428951 ns/op	 6731088 B/op	  107238 allocs/op
BenchmarkUnTracedPong5000QPS-4   	       1	1022341299 ns/op	29393688 B/op	  531123 allocs/op
PASS
ok  	go.opencensus.io/benchmarks/trace/grpc	10.086s
```

* REST ping-pong server
```shell
goos: darwin
goarch: amd64
pkg: go.opencensus.io/benchmarks/trace/rest
BenchmarkTracedPong2QPS-4        	       1	1001991307 ns/op	   76336 B/op	     355 allocs/op
BenchmarkTracedPong10QPS-4       	       1	1003484415 ns/op	  119472 B/op	     940 allocs/op
BenchmarkTracedPong100QPS-4      	       1	1002324398 ns/op	  608432 B/op	    7764 allocs/op
BenchmarkTracedPong1000QPS-4     	       1	1000503665 ns/op	 5487040 B/op	   74440 allocs/op
BenchmarkTracedPong5000QPS-4     	       1	1014700123 ns/op	27376104 B/op	  372173 allocs/op
BenchmarkUnTracedPong2QPS-4      	       1	1003877196 ns/op	   69040 B/op	     291 allocs/op
BenchmarkUnTracedPong10QPS-4     	       1	1004920891 ns/op	  113472 B/op	     886 allocs/op
BenchmarkUnTracedPong100QPS-4    	       1	1002632060 ns/op	  594160 B/op	    7473 allocs/op
BenchmarkUnTracedPong1000QPS-4   	       1	1000573682 ns/op	 5353152 B/op	   72370 allocs/op
BenchmarkUnTracedPong5000QPS-4   	       1	1026227316 ns/op	27064096 B/op	  363038 allocs/op
PASS
ok  	go.opencensus.io/benchmarks/trace/rest	10.086s
```

### CODE REVIEW

**Comment by odeke-em** [comment]
Alright, above are some benchmarks that I've started, PTAL.

File: benchmarks/trace/grpc/service.proto [hunk 1]
```diff
@@ -0,0 +1,16 @@
+syntax = "proto3";
+
+package benchmarks;
+
+service Pinger {
+  rpc PingUniTraced(Ping) returns (Pong) {}
```
Comment (rakyll) [inline benchmarks/trace/grpc/service.proto]: What is Uni?

PingTraced?

File: benchmarks/trace/grpc/service.proto [hunk 2]
```diff
@@ -0,0 +1,16 @@
+syntax = "proto3";
```
Comment (rakyll) [inline benchmarks/trace/grpc/service.proto]: Missing license header.

File: benchmarks/trace/rest/rest_test.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
+// Copyright 2017, OpenCensus Authors
```
Comment (rakyll) [inline benchmarks/trace/rest/rest_test.go]: Can't we have this under the trace package?

/trace/benchmark_test.go?

File: benchmarks/trace/grpc/benchmark_test.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
+// Copyright 2017, OpenCensus Authors
```
Comment (rakyll) [inline benchmarks/trace/grpc/benchmark_test.go]: This should be under an internal directory.

/internal/bechmarks/grpc/trace/...

File: benchmarks/trace/grpc/service.proto [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+package benchmarks;
+
+service Pinger {
+  rpc PingUniTraced(Ping) returns (Pong) {}
+  rpc PingUniUntraced(Ping) returns (Pong) {}
```
Comment (rakyll) [inline benchmarks/trace/grpc/service.proto]: You don't have to return a Pong? No one is handling the return value.

Comment (rakyll) [inline benchmarks/trace/grpc/service.proto]: You don't need two methods in the Pinger.

You need only a Ping and use the grpctrace package to enable tracing.

File: benchmarks/trace/grpc/benchmark_test.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+	"testing"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"go.opencensus.io/trace"
+	"google.golang.org/grpc"
+)
+
+func untracedPong(ctx context.Context, client PingerClient) (*Pong, error) {
```
Comment (rakyll) [inline benchmarks/trace/grpc/benchmark_test.go]: You are benchmarking your own instrumentation here, you should benchmark the grpctrace package instead.

Comment (odeke-em) [reply benchmarks/trace/grpc/benchmark_test.go]: Thanks. Let me fix that.

File: benchmarks/trace/grpc/service.proto (hunk 1 above)
Comment (odeke-em) [reply benchmarks/trace/grpc/service.proto]: Uni -- one direction, so not streaming.

**Comment by rakyll** [comment]
@odeke-em, are you still working on this PR? If not, can we close it?

File: benchmarks/trace/grpc/service.proto (hunk 1 above)
Comment (odeke-em) [reply benchmarks/trace/grpc/service.proto]: Yeah am going to remove this, and make it your suggestion.

**Comment by odeke-em** [comment]
Am going to close this and start it afresh.

**Pull request closed without merging**